docs/tdag.md

```diff
@@ -0,0 +1,105 @@
+# TaintDAG file format
+
+The Taint Directed Acyclic Graph (TaintDAG, TDAG) file format is tailored to facilitate fast recording of taint operations.
+
+It is a binary file format based on sparse files. It consists of a header and a number of subsections. The subsections stores information about:
+* taint sources, filename and offset information
+* taint output log, tainted values written to an output file
+* taint graph, the graph of how taint values are unioned from source taint and other unions.
+
+Because of the sparse layout it is very well suited for mmap'ing directly into the instrumented process address space.
+
+## Taint Sources, Unions and Ranges
+Whenever data is read from an input file, the data entering the program is labeled as source taint. Information about which file and at what offset is kept. This is the only way taint can originate in a program.
+
+Whenever an operation is performed where one or more operands are involved a union or range value is created. If the taint labels are adjacent (number wise), e.g. two consecutive source taint bytes, a range is created. If the labels are not adjacent a union of the two labels is created. Unions and ranges occupy the same amount of storage. The main difference is that a range can be extended to become a larger range.
+
+The main motivation for introducing ranges is to allow very efficient membership testing. If a taint label is already included in a range of taint values, the range can be reused. It is possible to walk the tree of unioned labels but it reqiures more computation.
+
+## Affects control flow
+In addition to being Source-, Union- or Range-Taint, each value is also marked if it affects control flow. The basic example is if a value `x` (having label `y`) read from file is compared against another value and a branch is taken based on the result. Whenever the comparison and and branch is executed taint with label `y` is marked as affecting control flow.
+
+Affects control flow propagates through unions and ranges meaning that if a value `v`, having label `w` where `w` is a union or range, affects control flow. Then each taint label represented by `w` is in turn marked as affecting control flow.
+
+## File format
+The general layout of the file is as follows:
+```
+[FileHdr][FileDescriptorMap][TDAGMapping][SinkLog]
+```
+
+### FileHdr
+```C
+struct FileHdr {
+  uint64_t fd_mapping_offset;
+  uint64_t fd_mapping_count;
+  uint64_t tdag_mapping_offset;
+  uint64_t tdag_mapping_size;
+  uint64_t sink_mapping_offset;
+  uint64_t sink_mapping_size;
+};
+```
+Each offset is relative to file (or memory mapping) start.
+
+### FDMappingHdr
+At `fd_mapping_offset` there are `fd_mapping_count` `FDMappingHdr` structures.
```

Comment:
"At `fd_mapping_offset` there is an array of `FDMappingHdr` structures. Length of the array is given by `fd_mapping_count`. " maybe?

Reply:
We should also mention that "FD" means file descriptor.

Reply:
It is mentioned just below, is that good enough?